---
title: "Lung Cancer Data Vignette"
author: "Perry Haaland"
date: "5/19/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

All of the explanations and background are taken from Chapter B of OODA Book. by  S. Marron. These come from an RNAseq study of lung cancer, and in particular was an early data set collected as part of The Cancer Genome Atlas, Tomczak et al. [Tomczak2015TCGA]. The focus here is on the gene CDKN2A, which has long been known to be involved in many types of cancer. 

The purpose of this vignette is to supplement the OODA Book by providing some examples of the analysis using the R environment. Consequently, we won't be repeating the full analysis or explanation provided in the book, but rather demonstrating how to recreate some of the highlights of the analysis.

Note that we are going to be using the collection of `tidyverse` packages and capabilities for the analysis. We also use the package `factoextra` for visualizing the results of the PCA analysis.

```{r, message=FALSE}
library(tidyverse)
library(factoextra)
```

## Reading and organizing the data

Early TCGA Data with RNAseq read depth curves is provided for 4 genes. The data are from Matt Wilkerson, Oct. 2011. The 4 genes are:

* PIK3CA
* CDKN2A
* P10
* STK11

There is one file in the `DerivedData` folder for each of the four genes. We are concerned with the `csv` files. Each file contains a matrix of data. The first column is the Genomic Coordinate, whichc can be considered a row name. The remaining columns are subjects or case names  (TCGA IDs). The data entries are read counts. The gene name is included in the name of the file.

For this example we read the CDKN2A gene file. We are going to call this data frame `cases_df` to denote that the cases or observational units are the columns. For some parts of the analysis it will be more convenient to reformat the data so that cases are in rows. When we do that below, we will call that data frame `rows_df`.

```{r}
cases_df <- read.csv("../DerivedData/LungCancer2011CDKN2A.csv",
                      header = TRUE) %>%
  as_tibble
```

As we see below, here are 180 cases, namely the number of columns -1. There are 1709 genes; namely, the number of rows.

```{r}
dim(cases_df)
```

A sample of the data as read are shown below:

```{r}
cases_df[1:5, 1:5]
```

The columns are the observed cases or samples. We can see that the names are extrememly long and make the dataset hard to view. For the purposes of simplicity of presentation, we are going to substitute simple case names to make the data frame easier to work with.

```{r}
names(cases_df)[-1] = paste0("Case", 1:(ncol(cases_df) - 1))
cases_df[1:5, 1:5]
```

For the purposes of graphing the data we are going to convert the Genomic Coordinates to relative positions. The relative positions are called Exonic NT number. We add that to the data frame.

```{r}
cases_df <- cases_df %>%
  mutate(NT.Number = 1:nrow(cases_df)) %>%
  select(Genomic.Coords, NT.Number, starts_with("Case"))
cases_df[1:5,1:5]
```


## Initial examination of the counts

The data values are the counts of RNA reads at each genomic location (row) and for each subject (column). We take a look at the distribution of counts. First we note that there is a very long right tail.

```{r}
x <- cases_df %>%
  select(-Genomic.Coords, -NT.Number) %>%
  as_vector %>%
  as.integer 
data.frame(x = x) %>%
  ggplot(aes(x = x)) +
    geom_histogram(bins = 100) +
  labs(x = "Read Count",
       y = "Frequency")

```

We readily see that most of the counts are 0 or close to 0.

```{r}
table(x)[1:10]
```

For the purposes of subsequent analysis, we will transform the counts using a log transformation. See below. 

### Plots of Read Counts versus Position

In order to make the first series of graphs, we first pivot the data table to make the data format more amendable to standard methods in R.

```{r}
long_df <- cases_df %>%
  pivot_longer(starts_with("Case"),
               names_to = "Case",
               values_to = "Count")
head(long_df)
```

In this first plot, we show each case with a different line. The y value, `Count` has been transformed by log10(y+1). The line colors are recycled and don't have any significance. We note that there are some regions of the genome that have very few or no reads.

```{r}
long_df %>%
  ggplot(aes(x= NT.Number, y = log10(Count+1), group = Case)) +
           geom_line(aes(color = Case)) +
  theme(legend.position = "none") +
  labs(x = "exonic NT number, not genomic position",
       y = "log10(RNA read depth + 1)",
       title = "Gene = CDKN2A")
  
```

## Principle Components Analysis

The plots above don't reveal any particularly interesting structure. The next step is to examine the results of Principle Components Analysis (PCA). To do this we are going to reformat the data so that cases are in rows.

```{r}
rows_df <- cases_df %>%
  select(-Genomic.Coords, -NT.Number) %>%
  t()
colnames(rows_df) <- cases_df$NT.Number
rows_df[1:5, 1:5]
```

Now we are ready to do the PCA. Note that we standardize the counts at each position as part of the PCA. The first visualization is the standard scree plot. Note that we use the log transformation as above.

```{r,message=FALSE}
pca.out <- prcomp(log10(rows_df + 1), scale = TRUE)
fviz_eig(pca.out)
```

Next we look at the scatter plot of the first two dimensions.

```{r}
fviz_pca_ind(pca.out,
             axes = c(1, 2),
             geom = "point")
```


```{r}
fviz_pca_ind(pca.out,
             axes = c(2, 3),
             geom = "point")
```
